spec/index.adoc

```diff
@@ -1674,6 +1674,96 @@ Any user can top-up any canister this way.
+=== Interface overview
+
+The following interface description, in https://github.com/dfinity/candid/blob/master/spec/Candid.md[Candid syntax], describes the available functionality. You can also link:{attachmentsdir}/bitcoin.did[download the file].
+----
+include::{example}bitcoin.did[]
+----
+
+The binary encoding of arguments and results are as per Candid specification.
+
+[#ic-bitcoin_get_utxos]
+=== IC method `get_utxos`
+
+Given a https://en.bitcoin.it/wiki/Base58Check_encoding[base58-encoded] address as part of the
+`GetUtxosRequest` , the function returns all unspent transaction outputs (UTXOs) associated with the
+provided address based on the current view of the Bitcoin blockchain available to the Bitcoin canister.
+
+If the call fails, e.g., because the address is malformed, a `GetUtxosError` variant is returned,
+indicating the reason for the failed call.
+
+The optional `min_confirmations` parameter can be used to limit the returned UTXOs to those with at
+least the provided number of confirmations, i.e., transactions with fewer than this number of
+confirmations are not considered.
+If this parameter is not used, the default value is 0.
+
+The optional `offset` parameter can be used to specify a starting offset in the list of UTXOs.
+Since a response will contain at most `TBD` `Utxo` records, this parameter is useful for addresses
+with many UTXOs.
+There is no guarantee that the set of UTXOs will remain unchanged between function calls with different
+offsets, i.e., every call will return the UTXOs starting from the provided offset based on the
+current view.
+If this parameter is not used, the default value is 0.
+
+[#ic-bitcoin_get_balance]
+=== IC method `get_balance`
+
+Given a https://en.bitcoin.it/wiki/Base58Check_encoding[base58-encoded] address as part of the
+`GetBalanceRequest` , the function returns the current balance of this address in `Satoshi` (10^8 Satoshi = 1 Bitcoin).
+
+If the call fails, e.g., because the address is malformed, a `GetBalanceError` variant is returned,
+indicating the reason for the failed call.
+
+The optional `min_confirmations` parameter can be used to limit the set of considered UTXOs
+for the calculation of the balance to those with at least the provided number of confirmations.
+
+Given an address and the optional `min_confirmations` parameter, `get_balance` iterates over
+all UTXOs returned by `get_utxos` for the same two parameters and returns the sum of
+UTXO values.
+Since the function is executed on the replica, there is no pagination limit, i.e., `get_balance`
+will iterate over all relevant UTXOs.
+
+[#ic-send_transaction]
+=== IC method `send_transaction`
+
+Given a `blob` of a Bitcoin transaction, several checks are performed:
+
+- The transaction is well formed.
+- The transaction only consumes unspent outputs.
+- All signatures are correct.
+- There is a positive transaction fee.
+- The transaction does not create dust, i.e., an output that holds a smaller Bitcoin amount
```

Comment:
How is this checked? There is no minimal transaction fee, or is there?

Reply:
The dust isn't about the transaction fee. It's about the value of the outputs.

Bitcoin core prevents transactions to be sent that have outputs below a certain value.

```
Bitcoin Core considers a transaction output to be dust, when its value is lower than the cost of spending it at the dustRelayFee rate. The default value for dustRelayFee is 3,000 sat/kvB
```

See https://bitcoin.stackexchange.com/questions/10986/what-is-meant-by-bitcoin-dust

Reply:
Right but doesn't
> ... than it costs to spend the Bitcoin in the output ...

relate to the transaction fee? What does it cost to spend the output if not the transaction fee? Can we put the definition of _dust_ we intend to use down in the specification here?

Reply:
The point is that an output is considered dust regardless of the fee that is paid for the transaction. In other words, even if there is a large fee for a transaction, an output can still be considered "dust" (namely when the Bitcoin amount in the output is less than it costs to spend it at the `dustRelayFee` rate).

However, as @ielashi once pointed out, it may be fine for us to ignore the dust problem entirely and let the users deal with the consequences of sending Bitcoin transactions that will likely not be added to blocks (we cannot make any promise that any transaction is added to a block anyway).

Reply:
Note: We decided to ignore the dust problem and remove this check.